tests/Unit/Domain/CoordinateMaps/Test_CubicScale.cpp

```diff
@@ -57,182 +57,198 @@ void cubic_scale_non_invertible(const double a0, const double b0,
-    const std::array<DataVector, deriv_order + 1> init_func_b{
+        CHECK(approx(scale_map_deserialized.hessian(point_xi, t, f_of_t_list)
+                         .get(0, 0, 0)) == time_time);
+        CHECK(approx(scale_map_deserialized.hessian(point_xi, t, f_of_t_list)
+                         .get(0, 1, 0)) == time_space);
+        CHECK(approx(scale_map_deserialized.hessian(point_xi, t, f_of_t_list)
+                         .get(0, 0, 1)) == time_space);
+        CHECK(approx(scale_map_deserialized.hessian(point_xi, t, f_of_t_list)
+                         .get(0, 1, 1)) == space_space);
+
+        // Check inequivalence operator
+        CHECK_FALSE(scale_map != scale_map);
+        CHECK_FALSE(scale_map_deserialized != scale_map_deserialized);
+
+        // Check serialization
+        CHECK(scale_map == scale_map_deserialized);
+        CHECK_FALSE(scale_map != scale_map_deserialized);
+
+        test_coordinate_map_argument_types(scale_map, point_xi, t, f_of_t_list);
+
+        t += dt;
+      }
+    };
+
+    std::array<DataVector, deriv_order + 1> init_func_a1{
+        {{1.0}, {0.0007}, {-0.004}}};
+    std::array<DataVector, deriv_order + 1> init_func_b1{
+        {{1.0}, {-0.001}, {0.003}}};
+    run_tests(init_func_a1, init_func_b1, 20.0, 19.2, 4.0);
+    // test again with inputs that explicitly result in multiple roots
+    std::array<DataVector, deriv_order + 1> init_func_a2{{{1.0}, {0.0}, {0.0}}};
+    std::array<DataVector, deriv_order + 1> init_func_b2{
         {{0.99}, {0.0}, {0.0}}};
-
-    FunctionsOfTime::PiecewisePolynomial<deriv_order> expansion_a(t,
-                                                                  init_func_a);
-    FunctionOfTime& expansion_a_base = expansion_a;
-
-    FunctionsOfTime::PiecewisePolynomial<deriv_order> expansion_b(t,
-                                                                  init_func_b);
-    FunctionOfTime& expansions_b_base = expansion_b;
-
-    const std::unordered_map<std::string, FunctionOfTime&> f_of_t_list = {
-        {"expansion_a", expansion_a_base}, {"expansion_b", expansions_b_base}};
-
-    const double outer_boundary = 20.0;
-    const CoordMapsTimeDependent::CubicScale scale_map(outer_boundary);
-    const std::array<double, 1> point_xi{{x0}};
-
-    while (t < final_time) {
-      const double a = expansion_a_base.func_and_deriv(t)[0][0];
-      const double b = expansions_b_base.func_and_deriv(t)[0][0];
-
-      const std::array<double, 1> mapped_point{
-          {point_xi[0] * (a + (b - a) * square(point_xi[0] / outer_boundary))}};
-
-      CHECK_ITERABLE_APPROX(scale_map(point_xi, t, f_of_t_list), mapped_point);
-      CHECK_ITERABLE_APPROX(
-          scale_map.inverse(mapped_point, t, f_of_t_list).get(), point_xi);
-      t += dt;
-    }
-  };
-
-  // test inverse at inner and outer boundary values
-  run_tests(0.0);
-  run_tests(20.0);
+    run_tests(init_func_a2, init_func_b2, 6.0, 5.0, 0.0);
+  }
+
+  // Test boundaries
+  {
```

Comment:
Again, please factor these scopes into separate functions. You should generally avoid functions that don't fit on your full laptop screen vertically. The longer the function the more difficult it is to understand, reason about, and catch bugs in.

Reply:
ping on this